pjmedia/src/pjmedia/event.c

```diff
@@ -107,15 +108,22 @@ static pj_status_t event_mgr_distribute_events(pjmedia_event_mgr *mgr,
             void *user_data = sub->user_data;
             pj_status_t status;
             
-            if (rls_lock)
+            if (rls_lock) {
+                /* To make sure that event unsubscription waits
+                 * until the callback completes.
+                 */
+                pj_mutex_lock(mgr->cb_mutex);
                 pj_mutex_unlock(mgr->mutex);
+            }
 
             status = (*cb)(ev, user_data);
             if (status != PJ_SUCCESS && err == PJ_SUCCESS)
 	        err = status;
 
-            if (rls_lock)
+            if (rls_lock) {
+                pj_mutex_unlock(mgr->cb_mutex);
                 pj_mutex_lock(mgr->mutex);
+            }
         }
 	sub = *next_sub;
```

Comment:
As `mgr->mutex` is released in the middle of list iteration, there is possibility that at this point the `*next_sub` is no longer valid, i.e: have unsubscribed and been removed from the list?

Reply:
I initially thought so too, but it turned out we already anticipated this. `next_sub` is actually a pointer to `mgr->th_next_sub` or `mgr->pub_next_sub`, so if the unsubscription removes this, it has to update either of those variables.

Reply:
ah okay, so `*next_sub` may be changed/updated by `unsubscribe()` while mutex is released.